File: VB/IncrementalSearch/Behavior/IncrementalSearchBehavior.vb
Imports Microsoft.VisualBasic
Imports DevExpress.Mvvm.UI.Interactivity
Imports DevExpress.Xpf.Editors
Imports DevExpress.Xpf.Grid
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports System.Windows
Imports System.Windows.Input
Imports System.Windows.Media
Imports System.Windows.Threading

Namespace IncrementalSearch
	Public Class IncrementalSearchBehavior
		Inherits Behavior(Of UIElement)
		Private Grid As GridControl
		Private TableView As TableView
		Private SearchingString As String
		Private PreviousColumn As GridColumn
		Private PreviousRowHandle As Integer
		Private Cell As FrameworkElement
		Private IsMatchRow As Boolean
		Private IsCustomFocus As Boolean

		Public Enum SearchParameter
			[Next]
			Pevious
		End Enum

		Protected Overrides Sub OnAttached()
			MyBase.OnAttached()
			Grid = TryCast(Me.AssociatedObject, GridControl)
			AddHandler Grid.Loaded, AddressOf Grid_Loaded
		End Sub

		Private Sub Grid_Loaded(ByVal sender As Object, ByVal e As RoutedEventArgs)
			TableView = TryCast(Grid.View, TableView)

			AddHandler Grid.PreviewTextInput, AddressOf Grid_PreviewTextInput
			AddHandler Grid.PreviewKeyDown, AddressOf Grid_PreviewKeyDown
			AddHandler Grid.CurrentColumnChanged, AddressOf Grid_CurrentColumnChanged
			AddHandler TableView.LayoutUpdated, AddressOf TableView_LayoutUpdated
			AddHandler TableView.FocusedRowHandleChanged, AddressOf TableView_FocusedRowHandleChanged
		End Sub

		Private Sub TableView_FocusedRowHandleChanged(ByVal sender As Object, ByVal e As FocusedRowHandleChangedEventArgs)
			If IsCustomFocus = False Then
				SearchingString = Nothing
				SetTextHighlight(Nothing, PreviousRowHandle, TryCast(Grid.CurrentColumn, GridColumn))

			Else
				IsCustomFocus = False
			End If
		End Sub

		Private Sub TableView_LayoutUpdated(ByVal sender As Object, ByVal e As EventArgs)
			TextHighlightUpdate()
		End Sub

		Private Sub Grid_CurrentColumnChanged(ByVal sender As Object, ByVal e As CurrentColumnChangedEventArgs)
			If PreviousColumn IsNot Nothing Then
				SetTextHighlight(Nothing, PreviousRowHandle, PreviousColumn)
				SearchingString = Nothing
				IsMatchRow = False
			End If
		End Sub

		Private Sub Grid_PreviewKeyDown(ByVal sender As Object, ByVal e As System.Windows.Input.KeyEventArgs)
			If Keyboard.Modifiers = ModifierKeys.Control AndAlso e.Key = Key.Down Then
				If IsMatchRow <> False AndAlso TableView.FocusedRowHandle + 1 < Grid.VisibleRowCount Then
					IsCustomFocus = True

					SetTextHighlight(Nothing, TableView.FocusedRowHandle, PreviousColumn)

					Dim handle = FindNextMatch(TableView.FocusedRowHandle, TryCast(Grid.CurrentColumn, GridColumn), SearchingString, SearchParameter.Next)
					TableView.FocusedRowHandle = handle

					PreviousRowHandle = TableView.FocusedRowHandle

					Cell = TableView.GetCellElementByRowHandleAndColumn(TableView.FocusedRowHandle, Grid.CurrentColumn)
					e.Handled = True
				End If
			ElseIf Keyboard.Modifiers = ModifierKeys.Control AndAlso e.Key = Key.Up Then
				If IsMatchRow <> False AndAlso TableView.FocusedRowHandle - 1 >= 0 Then
					IsCustomFocus = True

					SetTextHighlight(Nothing, TableView.FocusedRowHandle, PreviousColumn)

					Dim handle = FindNextMatch(TableView.FocusedRowHandle, TryCast(Grid.CurrentColumn, GridColumn), SearchingString, SearchParameter.Pevious)
					TableView.FocusedRowHandle = handle
					PreviousRowHandle = TableView.FocusedRowHandle

					Cell = TableView.GetCellElementByRowHandleAndColumn(TableView.FocusedRowHandle, Grid.CurrentColumn)
					e.Handled = True
				End If

			ElseIf e.Key = Key.Up OrElse e.Key = Key.Down Then
				SearchingString = Nothing
				SetTextHighlight(Nothing, TableView.FocusedRowHandle, PreviousColumn)
			End If
		End Sub

		Private Sub Grid_PreviewTextInput(ByVal sender As Object, ByVal e As System.Windows.Input.TextCompositionEventArgs)
			If e.OriginalSource.GetType() Is GetType(RowIndicator) OrElse e.OriginalSource.GetType() Is GetType(InplaceBaseEdit) Then
				Dim currentColumn = TryCast(Grid.CurrentColumn, GridColumn)

				If PreviousColumn IsNot currentColumn AndAlso PreviousColumn IsNot Nothing Then
					SearchingString = Nothing
				End If

				If e.Text Is Constants.vbBack AndAlso SearchingString IsNot Nothing AndAlso SearchingString.Length > 0 Then
					IsCustomFocus = True
					SearchingString = SearchingString.Remove(SearchingString.Length - 1)

				ElseIf e.Text IsNot Constants.vbBack AndAlso e.Text IsNot Constants.vbCr Then
					IsCustomFocus = True
					SearchingString &= e.Text
				End If

				TextSearch(currentColumn, SearchingString)

				e.Handled = True
			End If
		End Sub

		Public Sub TextSearch(ByVal selectedColumn As GridColumn, ByVal searchingString As String)
			IsMatchRow = False

			Dim matchRowHandle As Integer = 0

			If (Not String.IsNullOrEmpty(searchingString)) Then
				For i As Integer = 0 To Grid.VisibleRowCount - 1
					Dim handle = Grid.GetRowHandleByVisibleIndex(i)

					Dim cellValue = Grid.GetCellValue(handle, selectedColumn).ToString()

					If cellValue.StartsWith(searchingString, True, System.Globalization.CultureInfo.CurrentCulture) Then
						matchRowHandle = handle
						IsMatchRow = True
						Exit For
					End If
				Next i

				If IsMatchRow = True Then
					SetTextHighlight(Nothing, PreviousRowHandle, selectedColumn)

					TableView.FocusedRowHandle = matchRowHandle
					PreviousRowHandle = matchRowHandle

                    Dispatcher.BeginInvoke(New Action(Function() AnonymousMethod1(selectedColumn, searchingString, matchRowHandle)), DispatcherPriority.Render)

				Else
					Me.SearchingString = Me.SearchingString.Remove(Me.SearchingString.Length - 1)
				End If
			Else
				SetTextHighlight(Nothing, PreviousRowHandle, selectedColumn)
			End If

			PreviousColumn = selectedColumn
		End Sub
		
        Private Function AnonymousMethod1(ByVal selectedColumn As GridColumn, ByVal searchingString As String, ByVal matchRowHandle As Integer) As Boolean
            SetTextHighlight(searchingString, matchRowHandle, selectedColumn)
            Cell = TableView.GetCellElementByRowHandleAndColumn(matchRowHandle, selectedColumn)
            Return True
        End Function

		Public Function FindNextMatch(ByVal rowHandle As Integer, ByVal selectedColumn As GridColumn, ByVal searchingString As String, ByVal parameter As SearchParameter) As Integer
			Dim foundHandle As Integer = 0

			If (Not String.IsNullOrEmpty(searchingString)) Then
				Dim match As Boolean = False

				Dim currentHandle = Grid.GetRowVisibleIndexByHandle(rowHandle)

				If parameter = SearchParameter.Next Then
					For i As Integer = currentHandle + 1 To Grid.VisibleRowCount - 1
						Dim handle = Grid.GetRowHandleByVisibleIndex(i)

						Dim cellValue = Grid.GetCellValue(handle, selectedColumn).ToString()

						If cellValue.StartsWith(searchingString, True, System.Globalization.CultureInfo.CurrentCulture) Then
							foundHandle = handle
							match = True
							Exit For
						End If
					Next i

					If match = False Then
						foundHandle = currentHandle
					End If
				ElseIf parameter = SearchParameter.Pevious Then
					For i As Integer = currentHandle - 1 To 0 Step -1
						Dim handle = Grid.GetRowHandleByVisibleIndex(i)

						Dim cellValue = Grid.GetCellValue(handle, selectedColumn).ToString()

						If cellValue.StartsWith(searchingString, True, System.Globalization.CultureInfo.CurrentCulture) Then
							foundHandle = handle
							match = True
							Exit For
						End If
					Next i

					If match = False Then
						foundHandle = currentHandle
					End If
				End If
			End If

			Return foundHandle
		End Function

		Public Sub SetTextHighlight(ByVal valueString As Object, ByVal rowHandle As Integer, ByVal column As GridColumn)
			Dim cell = TryCast(TableView.GetCellElementByRowHandleAndColumn(rowHandle, column), CellEditor)

			If cell IsNot Nothing Then
				Dim editor = TryCast(VisualTreeHelper.GetChild(cell, 0), InplaceBaseEdit)
				Dim inplaceEditor = TryCast(editor, IInplaceBaseEdit)

				inplaceEditor.HighlightedText = TryCast(valueString, String)
				inplaceEditor.HighlightedTextCriteria = HighlightedTextCriteria.StartsWith
			End If
		End Sub

		Public Sub TextHighlightUpdate()
			Dim currentCell = TableView.GetCellElementByRowHandleAndColumn(TableView.FocusedRowHandle, Grid.CurrentColumn)

			If currentCell IsNot Nothing Then
				Cell = currentCell

				SetTextHighlight(SearchingString, TableView.FocusedRowHandle, TryCast(Grid.CurrentColumn, GridColumn))
			ElseIf Cell IsNot Nothing Then
				Dim editor = TryCast(VisualTreeHelper.GetChild(Cell, 0), InplaceBaseEdit)
				Dim inplaceEditor = TryCast(editor, IInplaceBaseEdit)

				inplaceEditor.HighlightedText = Nothing
			End If
		End Sub

		Protected Overrides Sub OnDetaching()
			MyBase.OnDetaching()
		End Sub
	End Class
End Namespace

File: CS/IncrementalSearch/Behavior/IncrementalSearchBehavior.cs
using DevExpress.Mvvm.UI.Interactivity;
using DevExpress.Xpf.Editors;
using DevExpress.Xpf.Grid;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows;
using System.Windows.Input;
using System.Windows.Media;
using System.Windows.Threading;

namespace IncrementalSearch
{
    public class IncrementalSearchBehavior : Behavior<UIElement>
    {
        private GridControl Grid;
        private TableView TableView;
        private string SearchingString;
        private GridColumn PreviousColumn;
        private int PreviousRowHandle;
        private FrameworkElement Cell;
        private bool IsMatchRow;
        private bool IsCustomFocus;

        public enum SearchParameter
        {
            Next,
            Pevious
        }

        protected override void OnAttached()
        {
            base.OnAttached();
            Grid = this.AssociatedObject as GridControl;
            Grid.Loaded += Grid_Loaded;
        }

        void Grid_Loaded(object sender, RoutedEventArgs e)
        {
            TableView = Grid.View as TableView;

            Grid.PreviewTextInput += Grid_PreviewTextInput;
            Grid.PreviewKeyDown += Grid_PreviewKeyDown;
            Grid.CurrentColumnChanged += Grid_CurrentColumnChanged;
            TableView.LayoutUpdated += TableView_LayoutUpdated;
            TableView.FocusedRowHandleChanged += TableView_FocusedRowHandleChanged;
        }

        void TableView_FocusedRowHandleChanged(object sender, FocusedRowHandleChangedEventArgs e)
        {
            if (IsCustomFocus == false)
            {
                SearchingString = null;
                SetTextHighlight(null, PreviousRowHandle, Grid.CurrentColumn as GridColumn);
            }

            else
            {
                IsCustomFocus = false;
            }
        }

        void TableView_LayoutUpdated(object sender, EventArgs e)
        {
            TextHighlightUpdate();
        }

        void Grid_CurrentColumnChanged(object sender, CurrentColumnChangedEventArgs e)
        {
            if (PreviousColumn != null)
            {
                SetTextHighlight(null, PreviousRowHandle, PreviousColumn);
                SearchingString = null;
                IsMatchRow = false;
            }
        }

        void Grid_PreviewKeyDown(object sender, System.Windows.Input.KeyEventArgs e)
        {
            if (Keyboard.Modifiers == ModifierKeys.Control && e.Key == Key.Down)
            {
                if (IsMatchRow != false && TableView.FocusedRowHandle + 1 < Grid.VisibleRowCount)
                {
                    IsCustomFocus = true;

                    SetTextHighlight(null, TableView.FocusedRowHandle, PreviousColumn);

                    var handle = FindNextMatch(TableView.FocusedRowHandle, Grid.CurrentColumn as GridColumn, SearchingString, SearchParameter.Next);
                    TableView.FocusedRowHandle = handle;

                    PreviousRowHandle = TableView.FocusedRowHandle;

                    Cell = TableView.GetCellElementByRowHandleAndColumn(TableView.FocusedRowHandle, Grid.CurrentColumn);
                    e.Handled = true;
                }
            }
            else if (Keyboard.Modifiers == ModifierKeys.Control && e.Key == Key.Up)
            {
                if (IsMatchRow != false && TableView.FocusedRowHandle - 1 >= 0)
                {
                    IsCustomFocus = true;

                    SetTextHighlight(null, TableView.FocusedRowHandle, PreviousColumn);

                    var handle = FindNextMatch(TableView.FocusedRowHandle, Grid.CurrentColumn as GridColumn, SearchingString, SearchParameter.Pevious);
                    TableView.FocusedRowHandle = handle;
                    PreviousRowHandle = TableView.FocusedRowHandle;

                    Cell = TableView.GetCellElementByRowHandleAndColumn(TableView.FocusedRowHandle, Grid.CurrentColumn);
                    e.Handled = true;
                }
            }

            else if (e.Key == Key.Up || e.Key == Key.Down)
            {
                SearchingString = null;
                SetTextHighlight(null, TableView.FocusedRowHandle, PreviousColumn);
            }
        }

        void Grid_PreviewTextInput(object sender, System.Windows.Input.TextCompositionEventArgs e)
        {
            if (e.OriginalSource.GetType() == typeof(RowIndicator) || e.OriginalSource.GetType() == typeof(InplaceBaseEdit))
            {
                var currentColumn = Grid.CurrentColumn as GridColumn;

                if (PreviousColumn != currentColumn && PreviousColumn != null)
                {
                    SearchingString = null;
                }

                if (e.Text == "\b" && SearchingString != null && SearchingString.Length > 0)
                {
                    IsCustomFocus = true;
                    SearchingString = SearchingString.Remove(SearchingString.Length - 1);
                }

                else if (e.Text != "\b" && e.Text != "\r")
                {
                    IsCustomFocus = true;
                    SearchingString += e.Text;
                }

                TextSearch(currentColumn, SearchingString);

                e.Handled = true;
            }
        }

        public void TextSearch(GridColumn selectedColumn, string searchingString)
        {
            IsMatchRow = false;

            int matchRowHandle = 0;

            if (!String.IsNullOrEmpty(searchingString))
            {
                for (int i = 0; i < Grid.VisibleRowCount; i++)
                {
                    var handle = Grid.GetRowHandleByVisibleIndex(i);

                    var cellValue = Grid.GetCellValue(handle, selectedColumn).ToString();

                    if (cellValue.StartsWith(searchingString, true, System.Globalization.CultureInfo.CurrentCulture))
                    {
                        matchRowHandle = handle;
                        IsMatchRow = true;
                        break;
                    }
                }

                if (IsMatchRow == true)
                {
                    SetTextHighlight(null, PreviousRowHandle, selectedColumn);

                    TableView.FocusedRowHandle = matchRowHandle;
                    PreviousRowHandle = matchRowHandle;

                    Dispatcher.BeginInvoke(new Action(() =>
                    {
                        SetTextHighlight(searchingString, matchRowHandle, selectedColumn);
                        Cell = TableView.GetCellElementByRowHandleAndColumn(matchRowHandle, selectedColumn);
                    }), DispatcherPriority.Render);
                }

                else
                {
                    SearchingString = SearchingString.Remove(SearchingString.Length - 1);
                }
            }
            else
            {
                SetTextHighlight(null, PreviousRowHandle, selectedColumn);
            }

            PreviousColumn = selectedColumn;
        }

        public int FindNextMatch(int rowHandle, GridColumn selectedColumn, string searchingString, SearchParameter parameter)
        {
            int foundHandle = 0;

            if (!String.IsNullOrEmpty(searchingString))
            {
                bool match = false;

                var currentHandle = Grid.GetRowVisibleIndexByHandle(rowHandle);

                if (parameter == SearchParameter.Next)
                {
                    for (int i = currentHandle + 1; i < Grid.VisibleRowCount; i++)
                    {
                        var handle = Grid.GetRowHandleByVisibleIndex(i);

                        var cellValue = Grid.GetCellValue(handle, selectedColumn).ToString();

                        if (cellValue.StartsWith(searchingString, true, System.Globalization.CultureInfo.CurrentCulture))
                        {
                            foundHandle = handle;
                            match = true;
                            break;
                        }
                    }

                    if (match == false)
                    {
                        foundHandle = currentHandle;
                    }
                }
                else if (parameter == SearchParameter.Pevious)
                {
                    for (int i = currentHandle - 1; i >= 0; i--)
                    {
                        var handle = Grid.GetRowHandleByVisibleIndex(i);

                        var cellValue = Grid.GetCellValue(handle, selectedColumn).ToString();

                        if (cellValue.StartsWith(searchingString, true, System.Globalization.CultureInfo.CurrentCulture))
                        {
                            foundHandle = handle;
                            match = true;
                            break;
                        }
                    }

                    if (match == false)
                    {
                        foundHandle = currentHandle;
                    }
                }
            }

            return foundHandle;
        }

        public void SetTextHighlight(object valueString, int rowHandle, GridColumn column)
        {
            var cell = TableView.GetCellElementByRowHandleAndColumn(rowHandle, column) as CellEditor;

            if (cell != null)
            {
                var editor = VisualTreeHelper.GetChild(cell, 0) as InplaceBaseEdit;
                var inplaceEditor = editor as IInplaceBaseEdit;

                inplaceEditor.HighlightedText = valueString as string;
                inplaceEditor.HighlightedTextCriteria = HighlightedTextCriteria.StartsWith;
            }
        }

        public void TextHighlightUpdate()
        {
            var currentCell = TableView.GetCellElementByRowHandleAndColumn(TableView.FocusedRowHandle, Grid.CurrentColumn);

            if (currentCell != null)
            {
                Cell = currentCell;

                SetTextHighlight(SearchingString, TableView.FocusedRowHandle, Grid.CurrentColumn as GridColumn);
            }
            else if (Cell != null)
            {
                var editor = VisualTreeHelper.GetChild(Cell, 0) as InplaceBaseEdit;
                var inplaceEditor = editor as IInplaceBaseEdit;

                inplaceEditor.HighlightedText = null;
            }
        }

        protected override void OnDetaching()
        {
            base.OnDetaching();
        }
    }
}
